debugrt/host/host.c

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/debugrt/host.h>
+#include <string.h>
+
+#ifdef _MSC_VER
+
+#include <Windows.h>
+
+static uint64_t _lock = 0;
+
+static void spin_lock()
+{
+    while (InterlockedCompareExchange(&_lock, 1, 0) == 1)
+    {
+        // TODO: Do we need to yield CPU here?
+        // Sleep(0);
+    }
+}
+
+static void spin_unlock()
+{
+    InterlockedExchange(&_lock, 0);
+}
+
+#elif defined __GNUC__
+
+static uint8_t _lock = 0;
+
+static void spin_lock()
+{
+    while (!__atomic_test_and_set(&_lock, __ATOMIC_SEQ_CST))
+    {
+        asm volatile("pause");
+    }
+}
+
+static void spin_unlock()
+{
+    __atomic_clear(&_lock, __ATOMIC_SEQ_CST);
+}
+
+/*
+** These functions are needed to notify the debugger. They should not be
+** optimized out even though they don't do anything in here.
+*/
+
+OE_NO_OPTIMIZE_BEGIN
+
+OE_EXPORT
+OE_NEVER_INLINE void oe_notify_gdb_enclave_creation(
```

Comment:
Why don't we define these 2 functions for both Linux and Windows, so we can get rid of the `#ifdef __linux__` at the call sites?

Reply:
> Why don't we define these 2 functions for both Linux and Windows, so we can get rid of the `#ifdef __linux__` at the call sites?

The Windows implementation will be different. In Windows we will notify the debugger via RaiseException instead of the debugger adding breakpoints to this function as done in Linux.